pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/InvalidSlf4jMessageFormatRule.java

```diff
@@ -183,10 +190,13 @@ private int getAmountOfExpectedArguments(final String variableName, final List<A
     }
 
     private int countPlaceholders(final AbstractJavaTypeNode node) {
-        int result = 0; // zero means, no placeholders, or we could not analyze the message parameter
-        ASTLiteral stringLiteral = node.getFirstDescendantOfType(ASTLiteral.class);
-        if (stringLiteral != null) {
-            result = StringUtils.countMatches(stringLiteral.getImage(), "{}");
+        // zero means, no placeholders, or we could not analyze the message parameter
+        int result = 0;
+        List<ASTLiteral> literals = node.findDescendantsOfType(ASTLiteral.class);
```

Comment:
should this rather be `findChildrenOfType`? That way we don't look at nested method calls that may mislead us?

Reply:
Yes, I took this shortcut because of the ASTAdditiveExpression. So, we *might* be prone to nested method calls, e.g. `log.trace(createMessage("Prefix Message {}", "Other part of the message {}"), arg0, arg1)`. Without knowing, what exactly `createMessage(...)` does, we might be right or wrong. We could skip it and e.g. check only grandchilds if ASTAdditiveExpression exists and otherwise just take the childs... WDYT?

Reply:
Maybe... we would have to check for either direct descendants and grand-children. I don't think there is any other construct we are interested on, is it?